# CMSE 202 Midterm (Section 004 - Fall 2025)

The goal of this midterm is to give you the opportunity to test out some of the skills that you've developed thus far this semester. In particular, you'll practice setting up a GitHub repository, committing and pushing repository changes, downloading data with command line tools, performing some data analysis, possibly using a new Python package, and writing a python class. You should find that you have all of the skills necessary to complete this exam with even just eight weeks of CMSE 202 under your belt!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, try doing Parts 0 and 1 first so that you have your repository set up and you download all necessary data files as they will be necessary to complete the assigned tasks. Let your instructor know right away if you have problems downloading the data!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time. If you have any questions during the exam, you may ask the instructor, the TA, or the LA privately: simply raise your hand and one of us will come over to you. 


**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However: The use of any person-to-person communication software or generative AI tools is absolutely not acceptable.** If you are seen accessing your email, using a collaborative cloud storage or document software (e.g. Slack, Google Documents), or generative AIs (e.g. ChatGPT), you will be at risk for receiving a zero on the exam. Google often does give an AI summary regardless of search, so to avoid misusing AI on the exam you should not copy-paste any of the questions from the exam directly into Google. 

**Keep your eyes on your screen!** Unfortunately, there isn't enough space in the room for everyone to sit at their own table so please keep your eyes on your own screen. This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

**Github Issues:** If at any point you encounter an issue with Github, such as a merge conflict, rather than spending your limited exam time trying to fix it, just write the code you would use to achieve the task in the notebook and move on. If you note the issue you encountered and have correct code in your notebook, you can still receive full credit. 

---
## Part 0: Academic integrity statement

Read the following statement and edit the markdown text to put your name in the statement. This is your commitment to doing your own authentic work on this exam.

> I, **Joel Chaple**, affirm that this exam represents my own authetic work, without the use of any unpermitted aids or generative AIs or person-to-person communication. I understand that this exam an an opportunity to showcase my own progress in developing and improving my computational skills and have done my best to demonstrate those skills.

## Part 1: Add to your Git repository to track your progress on your exam

Before you get too far along in the exam, you're going to add it to the `cmse202-f25-turnin` repository you created in class (or in Homework 1) so that you can track your progress on the exam and preserve the final version that you turn in. In order to do this you need to

1. Navigate to your `cmse202-f25-turnin` repository and create a new directory called `midterm`.
2. Move this notebook into that **new directory** in your repository, then **add it and commit it to your repository**.
3. Finally, to test that everything is working, "git push" the file so that it ends up in your GitHub repository.

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" respository (you should have done this in the previous homework assignment).

**Important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the noteobok, **none of your changes will be tracked**!

The file should now show up on your GitHub account in the `cmse202-f25-turnin` repository inside the `midterm` directory you just created.  Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. It can be good to get into a habit of committing your changes any time you make a significant modification.

&#9989; **Question 1.1 Put all the Git commands you used to add your midterm folder and file to your GitHub repository (2 points)**:

```bash
cd CMSE202/repositories/cmse202-f25-turnin/
mkdir midterm
```

&#9989; **Question 1.2 Put the code the instructor will need to clone your repository here (1 point)**:


```bash
git clone https://github.com/nasajoel2032/CMSE202-f25-turnin
```

&#9989; **Question 1.3 Describe in words what a branch is and what branches are used for (2 points)**:


A branch is an important and useful feature that is often used in git repositories. It can be thought of a tree log as being the 'main' branch, and the little branches to the side are git branches; you can do your work in the separate branches but at the end of the day if you do not like the progress or want to go back you can always snap the branch off and all your work is still saved in the main branch. Metaphors aside, branches allow you to work on your repo in such a way that you and other collaborators can see if they like what is on a repo and are able to make commits, without committing yet to the main branch. Of course, if you like what you see on the other branches, you can merge it with the main branch, and it now becomes a part of your main repository.

&#9989; **Question 1.4 (2 points)**: After completing each of Part 2, Part 3, Part 4, and Part 5 below, save your notebook, commit the changes to your Git repository using a descriptive commit message, and push the changes to GitHub. Note that there are large &#128721; **STOP** signs reminding you to do this after each part. **Note:** If Github issues are encountered, just write the code you would use to commit and push changes under each of the stop signs. 

## Part 2: Downloading and analyzing unfamiliar data



&#9989; **Question 2.1  Use the `curl` command to download the Iris dataset at**

`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data`

**to your working directory. Include the code you used to download the dataset below. (2 points)**:


``` bash
curl -O https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data
```

&#9989; **Question 2.2 Read in the dataset as a Pandas `DataFrame` and then display the first 10 rows. The dataset does not have a header, so be sure to set the header to be `None`. (2 points)**: 



``` bash
(base) chaplejo@io:~/CMSE202/repositories/cmse202-f25-turnin$ mv iris.data midterm/
```

In [11]:
# --- Import Common Libraries --- 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# --- Read iris.data using pandas --- 
iris = pd.read_csv('iris.data', header=None)
iris.head(10)

,0
0,5.1 3.5 1.4 0.2 Iris-setosa
1,4.9 3.0 1.4 0.2 Iris-setosa
2,4.7 3.2 1.3 0.2 Iris-setosa
3,4.6 3.1 1.5 0.2 Iris-setosa
4,5.0 3.6 1.4 0.2 Iris-setosa
5,5.4 3.9 1.7 0.4 Iris-setosa
6,4.6 3.4 1.4 0.3 Iris-setosa
7,5.0 3.4 1.5 0.2 Iris-setosa
8,4.4 2.9 1.4 0.2 Iris-setosa
9,4.9 3.1 1.5 0.1 Iris-setosa


&#9989; **Question 2.3 The columns of this dataset are Sepal Length, Sepal Width, Petal Length and Petal Width, and Iris Type. Add a header (column names with these names as strings) to the dataset to make it more descriptive. Then, compute and display the average, min, and max values in each numeric column, and display them. (2 points)**: 



In [34]:
names = ['Sepal Length (cm)', 'Sepal Width (cm)', 'Petal Length (cm)', 'Petal Width (cm)', 'Iris Type']
irishead = pd.read_csv('iris.data', header=None, names=names, sep=r'\s+')
irishead.describe()

,Sepal Length (cm),Sepal Width (cm),Petal Length (cm),Petal Width (cm)
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.054000,3.758667,1.198667
std,0.828066,0.433594,1.764420,0.763161
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


&#9989; **Question 2.4 Create a filtered `DataFrame` consisting only of "Iris-setosa" flowers with sepal length at least 5.3 (2 points)**: 

In [46]:
# To understand better
group = irishead.groupby('Iris Type')
group.describe()

Sepal Length (cm)                                              \
                            count   mean       std  min    25%  50%  75%  max   
Iris Type                                                                       
Iris-setosa                  50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8   
Iris-versicolor              50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0   
Iris-virginica               50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9   

                Sepal Width (cm)         ... Petal Length (cm)       \
                           count   mean  ...               75%  max   
Iris Type                                ...                          
Iris-setosa                 50.0  3.418  ...             1.575  1.9   
Iris-versicolor             50.0  2.770  ...             4.600  5.1   
Iris-virginica              50.0  2.974  ...             5.875  6.9   

                Petal Width (cm)                                            
                           count   mean       std  min  25%  50%  75%  max  
Iris Type                                                                   
Iris-setosa                 50.0  0.244  0.107210  0.1  0.2  0.2  0.3  0.6  
Iris-versicolor             50.0  1.326  0.197753  1.0  1.2  1.3  1.5  1.8  
Iris-virginica              50.0  2.026  0.274650  1.4  1.8  2.0  2.3  2.5  

[3 rows x 32 columns]

In [81]:
# Apply mask
setosamask = (irishead['Sepal Length (cm)'] >= 5.3) & (irishead['Petal Width (cm)'] <= 0.6) 
setosa = irishead[setosamask]
setosa # Verify results

,Sepal Length (cm),Sepal Width (cm),Petal Length (cm),Petal Width (cm)
count,11.000000,11.000000,11.000000,11.000000
mean,5.500000,3.809091,1.481818,0.281818
std,0.161245,0.317662,0.172152,0.098165
min,5.300000,3.400000,1.200000,0.200000
25%,5.400000,3.600000,1.350000,0.200000
50%,5.400000,3.800000,1.500000,0.200000
75%,5.600000,3.950000,1.600000,0.400000
max,5.800000,4.400000,1.700000,0.400000


&#9989; **Question 2.5 Using the filtered `DataFrame`, make histograms of petal lengths and petal widths (2 points)**:


In [80]:
x = setosa(['Petal Length (cm)'])
y = setosa(['Petal Width (cm)'])

TypeError: 'DataFrame' object is not callable

&#9989; **Question 2.6 How do the most frequent petal length and petal width in the filtered dataset compare to the average values in the full dataset? (2 points)**: 


In [82]:
setosa.describe()

,Sepal Length (cm),Sepal Width (cm),Petal Length (cm),Petal Width (cm)
count,11.000000,11.000000,11.000000,11.000000
mean,5.500000,3.809091,1.481818,0.281818
std,0.161245,0.317662,0.172152,0.098165
min,5.300000,3.400000,1.200000,0.200000
25%,5.400000,3.600000,1.350000,0.200000
50%,5.400000,3.800000,1.500000,0.200000
75%,5.600000,3.950000,1.600000,0.400000
max,5.800000,4.400000,1.700000,0.400000


In [83]:
irishead.describe()

,Sepal Length (cm),Sepal Width (cm),Petal Length (cm),Petal Width (cm)
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.054000,3.758667,1.198667
std,0.828066,0.433594,1.764420,0.763161
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


The petal lengths and petal widths in the filtered dataset have:
- minimum length:
  - 1.2 cm
- maximum length:
  - 1.7 cm
- mean length:
  - 1.48 cm
- minimum width:
  - 0.2 cm
- maximum width:
  - 0.4 cm
- mean width:
  - 0.28 cm
Whereas the petal lengths and petal widths in the unfiltered dataset have:
- minimum length:
  - 1 cm
- maximum length:
  - 6.9 cm
- mean length:
  - 3.78 cm
- minimum width:
  - 0.1 cm
- maximum width:
  -2.5 cm
- mean width:
  -1.20 cm

Overall, the trend shows that the filtered dataset, which looks for sepal lengths of setosa that are at least 5.3 cm, has much smaller average length and width values than the unfiltered dataset. This is because the unfiltered dataset has multiple species whereas the filtered only has setosa which is known for its smaller length and width values.

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 3: Working with a new Python package

In this section you will be exploring the [urllib3](https://pypi.org/project/urllib3/) Python package along with its [documentation](https://urllib3.readthedocs.io/en/stable/).


&#9989; **Question 3.1 Use pip install to install the package. Include the code you used to install it here (or, if you already have it installed, the code you would use to install it). (1 point):** 

In [84]:
pip install urllib3

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


&#9989; **Question 3.2 Following the documentation, make a GET request to https://httpbin.org/robots.txt and print out the data. (2 points)** 


In [6]:
# im getting 503 error code i can't access the txt file :(
import urllib3
x = urllib3.request('GET', 'https://httpbin.org/robots.txt')
x.status

200

In [7]:
x.data

b'User-agent: *\nDisallow: /deny\n'

&#9989; **Question 3.3 Make a GET request to the dataset used from the previous problem (https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data) and print out the received data (2 points):** 


In [11]:
getgit = urllib3.request('GET', 
'https://raw.githubusercontent.com/msu-cmse-courses/cmse202-supplemental-data/refs/heads/main/data/iris.data')
getgit.data

b'5.1 3.5 1.4 0.2 Iris-setosa\n4.9 3.0 1.4 0.2 Iris-setosa\n4.7 3.2 1.3 0.2 Iris-setosa\n4.6 3.1 1.5 0.2 Iris-setosa\n5.0 3.6 1.4 0.2 Iris-setosa\n5.4 3.9 1.7 0.4 Iris-setosa\n4.6 3.4 1.4 0.3 Iris-setosa\n5.0 3.4 1.5 0.2 Iris-setosa\n4.4 2.9 1.4 0.2 Iris-setosa\n4.9 3.1 1.5 0.1 Iris-setosa\n5.4 3.7 1.5 0.2 Iris-setosa\n4.8 3.4 1.6 0.2 Iris-setosa\n4.8 3.0 1.4 0.1 Iris-setosa\n4.3 3.0 1.1 0.1 Iris-setosa\n5.8 4.0 1.2 0.2 Iris-setosa\n5.7 4.4 1.5 0.4 Iris-setosa\n5.4 3.9 1.3 0.4 Iris-setosa\n5.1 3.5 1.4 0.3 Iris-setosa\n5.7 3.8 1.7 0.3 Iris-setosa\n5.1 3.8 1.5 0.3 Iris-setosa\n5.4 3.4 1.7 0.2 Iris-setosa\n5.1 3.7 1.5 0.4 Iris-setosa\n4.6 3.6 1.0 0.2 Iris-setosa\n5.1 3.3 1.7 0.5 Iris-setosa\n4.8 3.4 1.9 0.2 Iris-setosa\n5.0 3.0 1.6 0.2 Iris-setosa\n5.0 3.4 1.6 0.4 Iris-setosa\n5.2 3.5 1.5 0.2 Iris-setosa\n5.2 3.4 1.4 0.2 Iris-setosa\n4.7 3.2 1.6 0.2 Iris-setosa\n4.8 3.1 1.6 0.2 Iris-setosa\n5.4 3.4 1.5 0.4 Iris-setosa\n5.2 4.1 1.5 0.1 Iris-setosa\n5.5 4.2 1.4 0.2 Iris-setosa\n4.9 3.1 1.5 

&#9989; **Question 3.4 Write Python code to convert the received data into a list of lists. The numeric data (first four columns) should be `float`s, and the last column should be strings. Print out the first ten rows. (3 points):** 

*Hint 1: If you followed along with the urllib3 documentation, the received data should be of type bytes. You can convert this to a string using the `.decode("utf-8")` method for bytes.*

*Hint 2: The `.split` method of a string may be useful.*

In [ ]:
# Your code here.

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 4: Writing a simple Python class

In this section, we will be developing a Python class for a car.

&#9989; **Question 4.1 Complete the following class for a simple model of a car, referring to docstrings for instructions. (3 points):** 


In [ ]:
class Car:
    def __init__(self):
        """Initializes a car, an object with a fuel level between 0.0 and 1.0 that starts out as 1.0."""
        # Don't change this method!
        self.fuel = 1.0

    def get_fuel(self):
        """Returns the current fuel level (the fuel attribute)."""
        pass  # Your code here.

    def drive(self):
        """Reduces the fuel by half."""
        # Don't change this method!
        self.fuel = self.fuel / 2

    def add_fuel(self, fuel_to_add):
        """Adds fuel to the car.

        Args:
            fuel_to_add (float): The amount of fuel to add.

        Note: The car's fuel level cannot be greater than 1.0. If adding
        the fuel_to_add makes the fuel greater than 1.0, the fuel is set to 1.0.
        """
        pass  # Your code here.

&#9989; **Question 4.2 Create a `Car` object, and do the following: (3 points):** 

1. Call the `drive` method.
2. Call the `add_fuel` method to add 0.3 fuel.
3. Call the `drive` method.
4. Print out the fuel level.

In [ ]:
# Your code here.

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## Part 5: Writing Python code using different programming paradigms

In this section, you will show your ability to use the three different programming paradigms we learned in class (linear, functional, and object-oriented) on the following task:

Given a list of integers `numbers` and an integer `value`, add the `value` to each number in `numbers`, creating a new list.

Example:

```python
numbers = [1, 2, 3]
value = 5
# Expected output: [6, 7, 8]
```

**Your code must work for any list of integers (of any length) and any integer `value`.**

&#9989; **Question 5.1 Write code for the above task using linear programming (2 points):** 

In [ ]:
numbers = [1, 2, 3]
value = 5
# Your code here.

&#9989; **Question 5.2 Write code for the above task using functional programming, in particular by completing the following function (2 points):** 

In [ ]:
def add_value(numbers, value):
    pass  # Your code here.

&#9989; **Question 5.3 Write code for the above task using object-oriented programming, in particular by completing the following class (2 points):** 

In [ ]:
class ListOfNumbers:
    def __init__(self, numbers):
        """Initializes a list of numbers, storing the input numbers in an attribute called numbers.

        Args:
            numbers (list[int]): A list of numbers to store as an attribute in the class.
        """
        pass  # Your code here.

    def add_value(self, value):
        """Adds the value to every number in the numbers attribute, modifying numbers and returning nothing.

        Args:
            value (int): The value to add to each number.
        """
        pass  # Your code here.

&#9989; **Question 5.4 What are the advantages and disadvantages of each programming paradigm? (3 points):** 

Your answer here.

---
### &#128721; STOP (1 point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit  and push the changes to your Git repository.

---

## You're done! Congrats on finishing your CMSE 202 Midterm!

Make sure all of your changes to your repository are committed and pushed to GitHub. Also upload a copy of this notebook to the dropbox on D2L in case something went wrong with your repository or if you couldn't get the repository to work.